ManageIQ/manageiq-ui-classic
# Fix webpack tasks from `manageiq` with `manageiq-ui-classic` as a git-gem

Fixes many of the problems that don't translate when using `manageiq-ui-classic` as a git gem from `manageiq`, and not as a local copy with `override_gem` (which is how I assume https://github.com/ManageIQ/manageiq-ui-classic/pull/1132 was QA'd).

Changes the following:

* Avoids loading `jasmine` in the Rakefile (now not needed, but when we were calling `bundle exec rake` from the installation of `manageiq-ui-classic`, jasmine wouldn't be installed)
* Removes `Bundler.with_clean_env` in webpack tasks (also irrelevant, but was used when attempting to still shelling out to `rake`.  This allowed bundler to function properly since `git gems` from bundler don't have a seperate bundle, unlike when developing `manageiq-ui-classic` locally and hooking it )
* Uncomments the `dev_server` configs in `config/webpacker.yml`.  This is needed if `webpack:server` is to function, and if `manageiq-ui-classic` is installed as a git gem, this shouldn't have to be edited to do so (that said, this rake tasks doesn't really make sense to be used from `manageiq`, but ¯\\(°_o)/¯ )
* A chdir and system call using `bundle exec ...` was not required for `bin/webpack-dev-server` to function, so simplifed that in `webpack:server`
* Removes conditional around  `lib/tasks/manageiq/webpacker.rb` which makes it not define anything when loaded from `manageiq`.  The `webpack:paths` task is in fact needed for `webpack:compile` and `webpack:clobber` to function, which is defined in that file.
* Don't load `webpacker.rake` if it's tasks are already defined.  This was causing duplicate calls to `webpacker:compile` to be called, which was excessive.
* Make `webpack:paths` make use of a `file` tasks to avoid writing `config/webpack/paths.json` if it already exists
* Use `Rake::Task#invoke` to call `webpack:paths`, `webpacker:compile`, and `webpacker:clobber` from `webpack:compile` and `webpack:clobber`.  We still wrap everything in a `Dir.chdir`, but makes sure the `environment` task exists for the `manageiq` application.


Links
-----

* https://github.com/ManageIQ/manageiq-ui-classic/pull/1132
* https://github.com/ManageIQ/manageiq/issues/15748


Steps for Testing/QA
--------------------

You will want to run the QA steps in each of the following setups:

* From a local installation of `manageiq-ui-classic`
* From a local installation of `manageiq` with a `override_gem` pointing to `manageiq-ui-classic`
* From a local installation of `manageiq` with `manageiq-ui-classic` installed as a git gem

For the last setup, you can do the following to get these changes into into your copy of the get gem by running the following:

```console
$ bundle update manageiq-ui-classic
$ cd $(bundle show manageiq-ui-classic)
$ curl -L https://github.com/ManageIQ/manageiq-ui-classic/pull/1849.patch | patch -p 2
$ cd -
````

Make sure the following work:

* `rake update:ui`
* `rake webpack:server`


Bonus points if you can some how test this on an appliance.
